In [635]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
import statsmodels.api as sm

Подготовим данные для обучения модели

In [636]:
df = pd.read_csv('data\priora_data.csv')
pd.set_option('display.max_columns', None)
df

<>:1: SyntaxWarning: invalid escape sequence '\p'
<>:1: SyntaxWarning: invalid escape sequence '\p'
C:\Users\kpuck\AppData\Local\Temp\ipykernel_10896\3772101230.py:1: SyntaxWarning: invalid escape sequence '\p'
  df = pd.read_csv('data\priora_data.csv')


,title,price,price_score,Двигатель,Мощность,Коробка передач,Привод,Тип кузова,Цвет,Пробег,Владельцы,Руль,Поколение,Комплектация,pts,registration,gibdd,ur_face,period,search,limitations
0,"Продажа Лада Приора, 2010 год в Москве",380 000 ₽,нормальная цена,"Двигатель бензин, 1.6 л","Мощность 98 л.с., налог",Коробка передач механика,Привод передний,Тип кузова хэтчбек 5 дв.,Цвет черный,Пробег 130 000 км,NaN,Руль левый,Поколение 1 поколение,NaN,Характеристики совпадают с ПТС,5 записей о регистрации,Стоит на учете в ГИБДД,Не был во владении у юр. лица,Последний период владения больше 1 года,Не числится в розыске,Ограничений не обнаружено
1,"Продажа Лада Приора, 2009 год в Видном",550 000 ₽,высокая цена,"Двигатель бензин, 1.6 л","Мощность 98 л.с., налог",Коробка передач механика,Привод передний,Тип кузова хэтчбек 5 дв.,Цвет зеленый,Пробег 113 000 км,Владельцы 4 и более,Руль левый,Поколение 1 поколение,NaN,Характеристики совпадают с ПТС,6 записей о регистрации,Стоит на учете в ГИБДД,Не был во владении у юр. лица,Последний период владения больше 1 года,Не числится в розыске,Ограничений не обнаружено
2,"Продажа Лада Приора, 2008 год в Зеленограде",210 000 ₽,отличная цена,"Двигатель бензин, 1.6 л","Мощность 98 л.с., налог",Коробка передач механика,Привод передний,Тип кузова седан,Цвет зеленый,Пробег 233 000 км,Владельцы 4 и более,Руль левый,Поколение 1 поколение,Комплектация 1.6 MT Норма 21703-81-035,Характеристики совпадают с ПТС,9 записей о регистрации,Стоит на учете в ГИБДД,Не был во владении у юр. лица,Последний период владения больше 6 месяцев,Не числится в розыске,Ограничений не обнаружено
3,"Продажа Лада Приора, 2013 год в Пушкино",480 000 ₽,высокая цена,"Двигатель бензин, 1.6 л","Мощность 98 л.с., налог",Коробка передач механика,Привод передний,Тип кузова седан,Цвет синий,Пробег 179 000 км,Владельцы 3,Руль левый,Поколение 1 поколение,NaN,Характеристики совпадают с ПТС,3 записи о регистрации,Стоит на учете в ГИБДД,Не был во владении у юр. лица,Последний период владения больше 6 месяцев,Не числится в розыске,Ограничений не обнаружено
4,"Продажа Лада Приора, 2011 год в Кашире",250 000 ₽,без оценки,"Двигатель бензин, 1.6 л","Мощность 98 л.с., налог",Коробка передач механика,Привод передний,Тип кузова седан,Цвет белый,Пробег 237 834 км,Владельцы 4 и более,Руль левый,Поколение 1 поколение,Комплектация 1.6 MT Люкс 21703-02-018,Характеристики совпадают с ПТС,8 записей о регистрации,Не стоит на учете в ГИБДД,Был во владении у юр. лица,Последний период владения больше 2 лет,Не числится в розыске,Ограничений не обнаружено
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
315,"Продажа Лада Приора, 2009 год в Красноармейске",250 000 ₽,отличная цена,"Двигатель бензин, 1.6 л","Мощность 98 л.с., налог",Коробка передач механика,Привод передний,Тип кузова хэтчбек 5 дв.,Цвет серый,Пробег 215 000 км,Владельцы 3,Руль левый,Поколение 1 поколение,Комплектация 1.6 MT Норма 21723-01-011,Характеристики совпадают с ПТС,3 записи о регистрации,Стоит на учете в ГИБДД,Не был во владении у юр. лица,Последний период владения больше 5 лет,Не числится в розыске,Ограничений не обнаружено
316,"Продажа Лада Приора, 2009 год в Бронницах",365 000 ₽,нормальная цена,"Двигатель бензин, 1.6 л","Мощность 98 л.с., налог",Коробка передач механика,Привод передний,Тип кузова хэтчбек 5 дв.,Цвет зеленый,Пробег 140 000 км,Владельцы 4 и более,Руль левый,Поколение 1 поколение,NaN,Характеристики совпадают с ПТС,4 записи о регистрации,Стоит на учете в ГИБДД,Не был во владении у юр. лица,Последний период владения меньше 6 месяцев,Не числится в розыске,Ограничений не обнаружено
317,"Продажа Лада Приора, 2008 год в Москве",280 000 ₽,нормальная цена,"Двигатель бензин, 1.6 л","Мощность 98 л.с., налог",Коробка передач механика,Привод передний,Тип кузова седан,Цвет серебристый,Пробег 210 000 км,Владельцы 4 и более,Руль левый,Поколение 1 поколение,Комплектация 1.6 MT Люкс 21703-02-018,Характеристики совпадают с ПТС,8 записей о регистрации,Стоит на учете в 

In [637]:
df = df[df['price_score'].isna() == False]
df = df[df['price_score'] != 'без оценки']

In [638]:
df['price_score'] = df['price_score'].apply(lambda x: 0.25 if x == 'отличная цена' else 0.5 if x == 'хорошая цена' else 0.75 if x == 'нормальная цена' else 1)

In [639]:
df['price'] = df['price'].apply(lambda x: int(''.join(x.split()[:-1])))

In [640]:
df.drop(columns='Привод', inplace= True)

In [641]:
df.rename(columns = {
        'title': 'year',
        'Двигатель': 'engine',
        'Мощность': 'power',
        'Коробка передач': 'box',
        'Тип кузова': 'bodywork',
        'Цвет': 'color',
        'Пробег': 'mileage',
        'Владелец': 'owner',
        'Руль': 'wheel',
        'Поколение': 'generation',
        'Комплектация': 'equipment'
                    }, inplace = True)

In [642]:
df.drop(columns='Владельцы', inplace= True)

In [643]:
df['year'] = df['year'].apply(lambda x: int(x.split()[3]))
df['power'] = df['power'].apply(lambda x: int(x.split()[1]))
df['engine'] = df['engine'].apply(lambda x: float(x.split()[2]))

In [644]:
df['box'] = df['box'].apply(lambda x: str(x).split()[-1])
df['bodywork'] = df['bodywork'].apply(lambda x: ' '.join(x.split()[2:]))

In [645]:
df['color'] = df['color'].apply(lambda x: ' '.join(str(x).split()[1:]))
df['mileage']= df['mileage'].apply(lambda x: float(''.join(x.split()[1:-1])))
df.drop(columns= 'wheel', inplace=True)

In [646]:
df['generation'] = df['generation'].apply(lambda x: 2 if 'рестайлинг' in str(x) else 1)
df['pts'] = df['pts'].apply(lambda x: 0 if 'не' in str(x) else 1)
df['registration'] = df['registration'].apply(
    lambda x: int(str(x).split()[0]) if not pd.isna(x) else None
)
df['gibdd'] = df['gibdd'].apply(lambda x: 0 if 'Не' in str(x) else 1)
df['ur_face'] = df['ur_face'].apply(lambda x: 1 if 'Не' in str(x) else 0)

In [647]:
df.drop(columns= 'limitations', inplace=True)
df.drop(columns= 'search', inplace=True)
df.drop(columns='box', inplace=True)

In [648]:
df['period'] = df['period'].apply(lambda x: ' '.join(str(x).split()[-2:]))

In [649]:
def parse_period(x):
    if pd.isna(x):
        return None
    try:
        parts = str(x).split()
        value = float(parts[0])
        unit = parts[1]
        if 'месяц' in unit:
            return 0.5
        else:
            return int(value)
    except:
        return None

df['period'] = df['period'].apply(parse_period)

In [650]:
df['period'].value_counts()

period
0.5     112
1.0      48
2.0      22
3.0      11
5.0       7
4.0       3
9.0       3
7.0       3
17.0      2
12.0      2
6.0       2
8.0       1
14.0      1
10.0      1
11.0      1
Name: count, dtype: int64

In [651]:
df['equipment'] = df['equipment'].apply(lambda x: ' '.join(str(x).split()[3:4]))

In [652]:
df['equipment'] = df['equipment'].apply(lambda x: x.replace('', 'Норма') if len(x) == 0 else x)
df

,year,price,price_score,engine,power,bodywork,color,mileage,generation,equipment,pts,registration,gibdd,ur_face,period
0,2010,380000,0.75,1.6,98,хэтчбек 5 дв.,черный,130000.0,1,Норма,1,5.0,1,1,1.0
1,2009,550000,1.00,1.6,98,хэтчбек 5 дв.,зеленый,113000.0,1,Норма,1,6.0,1,1,1.0
2,2008,210000,0.25,1.6,98,седан,зеленый,233000.0,1,Норма,1,9.0,1,1,0.5
3,2013,480000,1.00,1.6,98,седан,синий,179000.0,1,Норма,1,3.0,1,1,0.5
6,2012,365000,0.75,1.6,98,седан,серый,214000.0,1,Норма,1,2.0,0,1,0.5
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
315,2009,250000,0.25,1.6,98,хэтчбек 5 дв.,серый,215000.0,1,Норма,1,3.0,1,1,5.0
316,2009,365000,0.75,1.6,98,хэтчбек 5 дв.,зеленый,140000.0,1,Норма,1,4.0,1,1,0.5
317,2008,280000,0.75,1.6,98,седан,серебристый,210000.0,1,Люкс,1,8.0,1,1,0.5
318,2012,520000,1.00,1.6,98,хэтчбек 5 дв.,черный,220000.0,1,Люкс,1,5.0,1,1,2.0


In [653]:
df = df[df['registration'].isna() == False]

In [654]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 219 entries, 0 to 319
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   year          219 non-null    int64  
 1   price         219 non-null    int64  
 2   price_score   219 non-null    float64
 3   engine        219 non-null    float64
 4   power         219 non-null    int64  
 5   bodywork      219 non-null    object 
 6   color         219 non-null    object 
 7   mileage       219 non-null    float64
 8   generation    219 non-null    int64  
 9   equipment     219 non-null    object 
 10  pts           219 non-null    int64  
 11  registration  219 non-null    float64
 12  gibdd         219 non-null    int64  
 13  ur_face       219 non-null    int64  
 14  period        219 non-null    float64
dtypes: float64(5), int64(7), object(3)
memory usage: 27.4+ KB


In [655]:
df.drop(columns= 'engine', inplace= True)

C:\Users\kpuck\AppData\Local\Temp\ipykernel_10896\2701557360.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(columns= 'engine', inplace= True)


Данные почти готовы для обучения модели. Для кодирования категориальных признаков используем Onehot Encoding

In [656]:
y = df['price_score']
df.drop(columns= 'price_score', inplace=True)

C:\Users\kpuck\AppData\Local\Temp\ipykernel_10896\3147782322.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(columns= 'price_score', inplace=True)


In [657]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_cols = ['bodywork', 'color', 'equipment']
X = df

preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ],
    remainder='passthrough'
)

X_transformed = preprocessor.fit_transform(X)

X = pd.DataFrame(X_transformed, columns=preprocessor.get_feature_names_out())
y = y.reset_index(drop=True)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42
)

model = sm.OLS(y_train, sm.add_constant(X_train))
results = model.fit()
display(results.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            price_score   R-squared:                       0.701
Model:                            OLS   Adj. R-squared:                  0.642
Method:                 Least Squares   F-statistic:                     11.88
Date:                Mon, 07 Apr 2025   Prob (F-statistic):           1.15e-22
Time:                        20:50:05   Log-Likelihood:                 112.05
No. Observations:                 153   AIC:                            -172.1
Df Residuals:                     127   BIC:                            -93.31
Df Model:                          25                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
const                          53.8043      9.405      5.721      0.000      35.194      72.415
cat__bodywork_седан            17.9707      3.132      5.738      0.000      11.773      24.168
cat__bodywork_универсал        17.8561      3.141      5.685      0.000      11.640      24.072
cat__bodywork_хэтчбек 3 дв.  6.816e-13   1.19e-13      5.720      0.000    4.46e-13    9.17e-13
cat__bodywork_хэтчбек 5 дв.    17.9775      3.132      5.740      0.000      11.780      24.175
cat__color_                     4.9691      0.857      5.795      0.000       3.272       6.666
cat__color_белый                4.9460      0.857      5.769      0.000       3.249       6.643
cat__color_бордовый             4.9090      0.861      5.701      0.000       3.205       6.613
cat__color_голубой              4.8501      0.858      5.656      0.000       3.153       6.547
cat__color_зеленый              4.8193      0.851      5.664      0.000       3.136       6.503
cat__color_золотистый        2.555e-14   4.47e-15      5.723      0.000    1.67e-14    3.44e-14
cat__color_коричневый           4.9689      0.862      5.767      0.000       3.264       6.674
cat__color_красный              4.8068      0.844      5.696      0.000       3.137       6.477
cat__color_серебристый          4.9258      0.856      5.752      0.000       3.231       6.620
cat__color_серый                4.8973      0.863      5.673      0.000       3.189       6.606
cat__color_синий                4.8364      0.861      5.618      0.000       3.133       6.540
cat__color_фиолетовый       -3.237e-17   2.71e-17     -1.194      0.235    -8.6e-17    2.13e-17
cat__color_черный               4.8755      0.856      5.694      0.000       3.181       6.570
cat__equipment_Comfort         13.3918      2.364      5.666      0.000       8.715      18.069
cat__equipment_Люкс            13.5435      2.346      5.773      0.000       8.901      18.186
cat__equipment_Норма           13.5220      2.343      5.772      0.000       8.886      18.158
cat__equipment_Спорт        -3.007e-17   2.95e-17     -1.020      0.310   -8.84e-17    2.83e-17
cat__equipment_Стандарт        13.3470      2.356      5.665      0.000       8.685      18.009
remainder__year                -0.0440      0.008     -5.604      0.000      -0.060      -0.028
remainder__price             1.926e-06   1.47e-07     13.088      0.000    1.64e-06    2.22e-06
remainder__power               -0.0158      0.005     -2.915      0.004      -0.026      -0.005
remainder__mileage           3.334e-07   1.76e-07      1.892      0.061   -1.53e-08    6.82e-07
remainder__generation          -0.2538      0.047     -5.362      0.000      -0.347      -0.160
remainder__pts                  0.1695      0.067      2.522      0.013       0.037       0.302
remainder__

Попробуем случайный лес, так как признаков много, и похоже, что есть нелинейная зависимость

In [682]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators=150,
                                max_depth= 20,
                                max_features= None, 
                                min_samples_leaf= 2, 
                                min_samples_split= 2, 
                                random_state=42)
model.fit(X, y)
print(model.score(X, y))

0.9386465309984515


In [683]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.37, random_state=42)

model.fit(X_train, y_train)

test_score = model.score(X_test, y_test)
print("r^2 на тестовой выборке:", test_score)

r^2 на тестовой выборке: 0.7318974368427638


Попробуем отбор признаков

In [672]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'n_estimators': [100, 200, 300, 500],
    'max_depth': [10, 20, 30, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['sqrt', 'log2', None]
}


rf = RandomForestRegressor(random_state=42)

grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=5, n_jobs=-1)

grid_search.fit(X_train, y_train)

print("Лучшие параметры:", grid_search.best_params_)

best_rf = grid_search.best_estimator_
print(f'R^2 на тестовых данных: {best_rf.score(X_test, y_test)}')

Лучшие параметры: {'max_depth': 20, 'max_features': None, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 300}
R^2 на тестовых данных: 0.733387761819549


73% на тестовой выборке достаточно неплохой результат, но можно попробовать применить и другие методы